feat: add guest user

skjnldsv opened the pull request:
## Features:
- Creates a new `getGuestUser` method that looks similar to a normal user
- Add a way to monitor displayName change to so whatever you want (reload an avatar, or change other stuff...)


## Code snippet example
```ts
// Watching for displayname change in vue component
data() {
	return {
		displayName: getGuestUser().displayName,
	}
},

mounted() {
	subscribe('user:info:changed', (guest) => { this.displayName = guest.displayName })
},

```

susnux commented on lib/guest.ts:
```diff
@@ -3,20 +3,66 @@
+
+}
+
+let currentUser: GuestUser | null | undefined
```
```suggestion
let currentUser: GuestUser | undefined
```

When could this be null?

susnux commented on lib/guest.ts:
```diff
@@ -3,20 +3,66 @@
+	type: 'updateDisplayName'
+}
+
+class GuestUser extends TypedEventTarget<{ updateDisplayName: UpdateGuestDisplayName }> {
```
Maybe also add the inheritance of `NextcloudUser` if you already implement `isAdmin`?

```suggestion
class GuestUser extends TypedEventTarget<{ updateDisplayName: UpdateGuestDisplayName }> implements NextcloudUser {
```

skjnldsv commented:
> Maybe it makes things easier to work with (with current design of nextcloud-auth) when having a global event like `auth:guest:display-name-changed` (or any other wording 😅 ).

`user:properties:changed` ?
So we can also use it for the `NextcloudUser` ?

That would differenciate for other like
- `user:preferences:changed`
- `user:config:changed`
or something ?

susnux commented on lib/guest.ts:
```diff
@@ -3,20 +3,68 @@
+/**
+ * Get the currently Guest user or null if not logged in
+ */
+export function getGuestUser(): GuestUser {
```
Either
```suggestion
export function getGuestUser(): NextcloudUser {
```

Or we would need to also export the type of GuestUser.

susnux reviewed (approved):
Looks good but we need the type adjusted or exported
https://github.com/nextcloud-libraries/nextcloud-auth/pull/795/files#r2077161323

susnux commented:
Damn did not saw the auto merge 😮‍💨

skjnldsv commented on lib/guest.ts:
```diff
@@ -3,20 +3,68 @@
 
+class GuestUser implements NextcloudUser {
+
+	private _displayName: string | null
```
I thought # was dropped in the end? :thinking:

susnux replied:
No they are standardized in es2019:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Private_properties

Its more like `protected` but enforced by runtime, meaning you cannot access the property from outside the class.

Whereas the typescript `private`, `protected`, and `readonly` are just types which get stripped away:

![Screenshot 2025-05-07 at 16-35-39 TypeScript TS Playground - An online editor for exploring TypeScript and JavaScript](https://github.com/user-attachments/assets/14dcaa63-286d-4bc2-b050-217c6159c068)

Meaning you could still access those private fields, while `#something` will throw an exception:
![grafik](https://github.com/user-attachments/assets/27d84efa-d819-4890-81f4-b9c4f24b3107)

---

In the end its not really important, both have good reasons to use or not to use, it depends :)

skjnldsv replied:
No, absolutely! 
But I remember using the `#` a while back, when it was drafted and about to be accepted.
But I also remember that we rolled back and stopped using it :thinking:

susnux replied:
there are some places using it, but it really is not that important.
If you mess with private fields you better know what you are doing, so its your risk.

Meaning its just a matter of taste if you prefer Typescript sugar or Javascript syntax.
E.g. a benefit of Typescript sugar is that you can unit test it (while also there are some guidelines that discurage you from testing private properties, but thats another topic^^).

skjnldsv commented:
> Damn did not saw the auto merge 😮‍💨

https://github.com/nextcloud-libraries/nextcloud-auth/pull/799
